gpu_bdb/queries/q03/gpu_bdb_query_03_dask_sql.py

```diff
@@ -0,0 +1,232 @@
+q03_days_in_sec_before_purchase = 864000
+q03_views_before_purchase = 5
+q03_purchased_item_IN = 10001
+# --see q1 for categories
+q03_purchased_item_category_IN = "2,3"
+q03_limit = 100
+
+
+@cuda.jit
+def find_items_viewed_before_purchase_kernel(
+    relevant_idx_col, user_col, timestamp_col, item_col, out_col, N
+):
+    """
+    Find the past N items viewed after a relevant purchase was made,
+    as defined by the configuration of this query.
+    """
+    i = cuda.grid(1)
+    relevant_item = q03_purchased_item_IN
+
+    if i < (relevant_idx_col.size):  # boundary guard
+        # every relevant row gets N rows in the output, so we need to map the indexes
+        # back into their position in the original array
+        orig_idx = relevant_idx_col[i]
+        current_user = user_col[orig_idx]
+
+        # look at the previous N clicks (assume sorted descending)
+        rows_to_check = N
+        remaining_rows = user_col.size - orig_idx
+
+        if remaining_rows <= rows_to_check:
+            rows_to_check = remaining_rows - 1
+
+        for k in range(1, rows_to_check + 1):
+            if current_user != user_col[orig_idx + k]:
+                out_col[i * N + k - 1] = 0
+
+            # only checking relevant purchases via the relevant_idx_col
+            elif (timestamp_col[orig_idx + k] <= timestamp_col[orig_idx]) & (
+                timestamp_col[orig_idx + k]
+                >= (timestamp_col[orig_idx] - q03_days_in_sec_before_purchase)
+            ):
+                out_col[i * N + k - 1] = item_col[orig_idx + k]
+            else:
+                out_col[i * N + k - 1] = 0
+
+def load_q03():
+    import importlib, types
+
+    fn = os.path.join(os.getcwd(), "gpu_bdb_query_03_dask_sql.py")
+    if not os.path.isfile(fn):
+        fn = os.path.join(os.getcwd(), "queries/q03/gpu_bdb_query_03_dask_sql.py")
+
+    loader = importlib.machinery.SourceFileLoader("03", fn)
+    mod = types.ModuleType(loader.name)
+    loader.exec_module(mod)
+    sys.modules[loader.name] = mod
+    return mod.main
+
+def apply_find_items_viewed(df, item_mappings):
+    import cudf
+
+    # need to sort descending to ensure that the
+    # next N rows are the previous N clicks
+    df = df.sort_values(
+        by=["wcs_user_sk", "tstamp", "wcs_sales_sk", "wcs_item_sk"],
+        ascending=[False, False, False, False],
+    )
+    df.reset_index(drop=True, inplace=True)
+    df["relevant_flag"] = (df.wcs_sales_sk != 0) & (
+        df.wcs_item_sk == q03_purchased_item_IN
+    )
+    df["relevant_idx_pos"] = df.index.to_series()
+    df.reset_index(drop=True, inplace=True)
+    # only allocate output for the relevant rows
+    sample = df.loc[df.relevant_flag == True]
+    sample.reset_index(drop=True, inplace=True)
+
+    N = q03_views_before_purchase
+    size = len(sample)
+
+    # we know this can be int32, since it's going to contain item_sks
+    out_arr = cuda.device_array(size * N, dtype=df["wcs_item_sk"].dtype)
+
+    load_q03()
```

Comment:
Confused about this

Reply:
This relates to the above comment that Ayush replied to

Reply:
Thanks.